pkg/server/admin.go

```diff
@@ -1397,6 +1397,43 @@ func (s *adminServer) checkReadinessForHealthCheck() error {
 		return status.Errorf(codes.Unavailable, "node is shutting down")
 	}
 
+	if !s.server.sqlServer.acceptingClients.Get() {
+		return status.Errorf(codes.Unavailable, "node is not accepting SQL clients")
+	}
+
+	// Also check the availability of the SQL service.
+	// We do need a timeout on the query, to ensure the health probe
+	// returns a result within a reasonable timeframe, and that
+	// unavailable ranges translate to a health failure.
+	timeoutCtx, cancel := context.WithTimeout(ctx, 1*time.Second)
+	defer cancel()
+	if err := s.server.RunLocalSQL(timeoutCtx, func(ctx context.Context, ie *sql.InternalExecutor) error {
+		_, err := ie.Exec(ctx, "sql-health", nil, `
+SELECT
+  (SELECT count(*) FROM (TABLE system.public.namespace      LIMIT 1))+
+  (SELECT count(*) FROM (TABLE system.public.users          LIMIT 1))+
+  (SELECT count(*) FROM (TABLE system.public.role_options   LIMIT 1))+
+  (SELECT count(*) FROM (TABLE system.public.role_members   LIMIT 1))+
+  (SELECT count(*) FROM (TABLE system.public.jobs           LIMIT 1))+
+  (SELECT count(*) FROM (TABLE system.public.sqlliveness    LIMIT 1))+
+  (SELECT count(*) FROM (TABLE system.public.scheduled_jobs LIMIT 1))+
+  (SELECT count(*) FROM (TABLE system.public.lease          LIMIT 1))
+`+
+			// We use AS OF SYSTEM TIME so that the health query does not
+			// need to create a hotspot at these tables' leaseholder.
+			// However, in order to use AOST we also need a FROM clause.
+			// We use generate_series(1,1) which is a dummy source guaranteed
+			// to return one row.
+			`FROM generate_series(1,1)  AS OF SYSTEM TIME '-5s'
+`)
+		return err
+	}); err != nil {
+		// TODO(knz): When always-on tracing is available via SQL, we may
```

Comment:
You don't need always-on tracing here - you can just trace it verbosely, period. The main reason not to is that we shouldn't verbosely trace anywhere until we have proper memory accounting, see https://github.com/cockroachdb/cockroach/projects/57#card-53177248